### DISCUSSION TITLE
MQTT Subscribe2 command for true/false value

### DISCUSSION THREAD

**Discussion by cmangla** [arendst/Tasmota#19100] [Q&A] [answered]
I have a rule that tries to subscribe to an MQTT value like so:
```
ON mqtt#connected DO Subscribe2 Occupied, zigbee2mqtt/ikea-motion, occupancy ENDON
```
In the log, I get:
```
 RUL: MQTT#CONNECTED performs "Subscribe2 Occupied, zigbee2mqtt/ikea-motion, occupancy"
```

The `occupancy` item is a json key whose value is true/false. The `Subscribe2` command seems to dislike it and just gives me `RESULT = {"Command":"Unknown"}` in the log when the MQTT message is published.

Q1: Is subscription to true/false json keys supported?

Q2: Where can I find the source code for the implementation of `Subscribe2`? I searched around but failed.

**Comment by sfromis** [comment]
For `Subscribe` and hence `Subscribe2` (same with index 2), to work, you need a build including this feature.
https://tasmota.github.io/docs/MQTT/#subscribeunsubscribe

Yes, it works with true/false, parsed as a string.

The subscribe command is implemented here:
https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_10_rules.ino#L1222-L1309

When it comes to your "unknown" command, do check log details with weblog 3 activated. And make sure to not be using an old Tasmota build.

**Comment by cmangla** [comment]
Many thanks for your reply @sfromis .

> For `Subscribe` and hence `Subscribe2` (same with index 2), to work, you need a build including this feature. https://tasmota.github.io/docs/MQTT/#subscribeunsubscribe

Yes, I did compile it in.

> Yes, it works with true/false, parsed as a string.

This is how I wrote the rule:
```
ON Event#Occupied=true DO if (Var1 == 1) RuleTimer1 3600; Var1 0; Power1 off ENDIF ENDON
```

Is my syntax correct? Do I need to quote `true`?

I will check and report the log with weblog 3 activated. I am running Tasmota 12.5.0

**Reply by cmangla** [reply]
Here is what I get in the log with weblog 3:
```
11:14:14.835 CMD: Grp 0, Cmd 'IKEA-MOTION', Idx 1, Len 259, Pld -99, Data '{"battery":87,"illuminance_above_threshold":false,"linkquality":51,"occupancy":false,"requested_brightness_level":76,"requested_brightness_percent":30,"update":{"installed_version":537011747,"latest_version":537011747,"state":"idle"},"update_available":false}'
11:14:14.841 MQT: stat/tasmota_1C947C/RESULT = {"Command":"Unknown"}
```

**Reply by sfromis** [reply]
No, rules do not use the concept of quoting strings. Note that if your build does not include the optional rules feature of expressions, `IF` is not available.

**Reply by sfromis** [reply]
Better have rules always matching the payload, otherwise Tasmota may treat the message as a command to be executed.

**Reply by cmangla** [reply]
> Note that if your build does not include the optional rules feature of expressions, `IF` is not available.

I've compiled with:
```
#define USE_EXPRESSION
#define SUPPORT_IF_STATEMENT
```

**Reply by cmangla** [reply]
I added to the rule for the `false` value:
```
	ON Event#Occupied=true DO if (Var1 == 1) RuleTimer1 3600; Var1 0; Power1 off ENDIF ENDON
	ON Event#Occupied=false DO ENDON
```
but I'm still getting:
```
11:42:56.058 CMD: Grp 0, Cmd 'IKEA-MOTION', Idx 1, Len 259, Pld -99, Data '{"battery":87,"illuminance_above_threshold":false,"linkquality":105,"occupancy":true,"requested_brightness_level":76,"requested_brightness_percent":30,"update":{"installed_version":537011747,"latest_version":537011747,"state":"idle"},"update_available":false}'
11:42:56.064 MQT: stat/tasmota_1C947C/RESULT = {"Command":"Unknown"}
```

**Reply by barbudor** [reply]
You are clearly trying to have use a command 'IKEA-MOTION' that obviously doesn't exist

**Reply by barbudor** [reply]
Something is wrong in your subscribe otherwise tasmota would not try to understand IKEA-MOTION as a command

Tasmota doesn't seems to match the incoming topic with the event

**Reply by sfromis** [reply]
Are your intended rules firing?

**Reply by barbudor** [reply]
Looking back at one subcription case I have for a thermostat based on Sonoff Basic, getting the current temperature from a Xiaomi BLE sensor
Looking to get the temperature from
```
tele/blegateway/room3 = {"Time":"2023-07-12T13:55:57+02:00","alias":"room3","mac":"a4c1384c3eff","Temperature":25.0,"Humidity":49.0,"DewPoint":13.5,"Btn":1,"Battery":83,"RSSI":-81,"TempUnit":"C"}
```
using 
```
subscribe settemp,tele/blegateway/room3,Temperature
```
and rule
```
on event#settemp do var1 %value% endon
```

I would suggest you reduce your rules to something simpler until you see really the event being triggered

**Comment by cmangla** [comment]
@barbudor @sfromis thanks for your replies. I have some more details.

Here is my rule in its entirety:
```
Rule2
	ON mqtt#connected DO Subscribe2 Occupied, zigbee2mqtt/ikea-motion, occupancy ENDON
	ON Event#Occupied=true DO if (Var1 == 1) RuleTimer1 3600; Var1 0; Power1 off ENDIF ENDON
	ON Event#Occupied=false DO ENDON
	ON Rules#Timer=1 DO if (TIME < 1200 AND Var1 == 0) Var1 1; Power1 on ENDIF ENDON
```

I have an IKEA motion sensor, that is connected to zigbee2mqtt which sends events to my MQTT server. Tasmota is installed on a Sonoff S26 (must like Sonoff basic) which needs to trigger a switch based on it, so that is where I'm trying to setup the rules.

@sfromis It seems the rule is indeed firing. It seems the `Subscribe2` command does get executed on `mqtt#connect`, which I check the results of by running just `Subscribe2` command in the console:
```
08:47:11.685 CMD: subscribe2
08:47:11.688 SRC: WebConsole from 192.168.1.141
08:47:11.690 CMD: Grp 0, Cmd 'SUBSCRIBE', Idx 2, Len 0, Pld -99, Data ''
08:47:11.696 MQT: stat/tasmota_1C947C/RESULT = {"Subscribe":"OCCUPIED,zigbee2mqtt/ikea-motion,occupancy; "}
```
The mqtt subscription is to a motion sensor, so when I walk in front of it, I get this in the Tasmota console:
```
08:50:12.722 SRC: MQTT
08:50:12.725 CMD: Grp 0, Cmd 'IKEA-MOTION', Idx 1, Len 258, Pld -99, Data '{"battery":87,"illuminance_above_threshold":false,"linkquality":80,"occupancy":true,"requested_brightness_level":76,"requested_brightness_percent":30,"update":{"installed_version":537011747,"latest_version":537011747,"state":"idle"},"update_available":false}'
08:50:12.730 MQT: stat/tasmota_1C947C/RESULT = {"Command":"Unknown"}
```
My guess is the `08:50:12.725 CMD` line is debug for what's read from MQTT, and you can see all the data is read correctly from MQTT. I am also guessing that the `08:50:12.730 MQT` line is the result of the rule-command executing.

@barbudor I'm not sure why the line `08:50:12.725 CMD` shows the command as being `IKEA-MOTION`, but it appears when the MQTT message is published and so I'm guessing it is just debut output showing data received? It does have all the correct data in it in the debug.

**Reply by sfromis** [reply]
Nothing in the redacted details you posted indicates that the intended rules are firing, merely suggests that the mqtt#connected one did. Anytime a rule trigger fires (including the Event ones) you should get message about this. `RUL: .... performs "..."`

The message with CMD being "IKEA-MOTION" indicates that the message was processed as a command, and it is not merely a debug message about what was received.

**Comment by cmangla** [comment]
@barbudor Coincidently, I also have a setup with a Xiaomi BLE sensor and a Sonoff Basic running Tasmota. The sensor BLE data is read by openmqttgateway which generates MQTT messages. Those are read by rules in Tasmota on Sonoff Basic using `Subscribe2` command, similar to the one I'm using above. That all works just fine for me.

**Comment by cmangla** [comment]
On @barbudor suggestion, I simplified my rule to:
```
Rule2
	ON mqtt#connected DO Subscribe2 Occupied, zigbee2mqtt/ikea-motion, occupancy ENDON
	ON Event#Occupied DO Var13 %value%  ENDON
	ON Event#GetVar13 DO Var13 ENDON
```

But, when the MQTT event occurs I still get:
```
09:58:26.406 SRC: MQTT
09:58:26.409 CMD: Grp 0, Cmd 'IKEA-MOTION', Idx 1, Len 258, Pld -99, Data '{"battery":87,"illuminance_above_threshold":false,"linkquality":83,"occupancy":true,"requested_brightness_level":76,"requested_brightness_percent":30,"update":{"installed_version":537011747,"latest_version":537011747,"state":"idle"},"update_available":false}'
09:58:26.414 MQT: stat/tasmota_1C947C/RESULT = {"Command":"Unknown"}
```
and nothing seems to have been saved in `Var13`, which I check using the `event GetVar13` command in the console:
```
09:58:47.102 CMD: event getvar13
09:58:47.104 SRC: WebConsole from 192.168.1.141
09:58:47.106 CMD: Grp 0, Cmd 'EVENT', Idx 1, Len 8, Pld -99, Data 'getvar13'
09:58:47.112 MQT: stat/tasmota_1C947C/RESULT = {"Event":"Done"}
09:58:47.160 RUL: EVENT#GETVAR13 performs "Var13"
09:58:47.162 SRC: Rule
09:58:47.165 CMD: Grp 0, Cmd 'VAR', Idx 13, Len 0, Pld -99, Data ''
09:58:47.170 MQT: stat/tasmota_1C947C/RESULT = {"Var13":""}
```

**Reply by sfromis** [reply]
Do look for the messages about rules firing, instead of trying to infer it from what you expect to be set.

**Reply by cmangla** [reply]
Thanks for the clarification regarding the `RUL:` logs @sfromis . Searching through my syslogs reveals that I do see the rule trigger for mqtt connection:
```
sw2 ESP-RUL: MQTT#CONNECTED performs "Subscribe2 Occupied, zigbee2mqtt/ikea-motion, occupancy"
```
and also for when I manually call `event GetVar13` in the console:
```
sw2 ESP-RUL: EVENT#GETVAR13 performs "Var13"
```
but *not* for the event related to receiving the MQTT message.

**Comment by cmangla** [comment]
@sfromis @barbudor I made some progress with debugging this. I changed this rule, which doesn't work:
```
Rule2
	ON mqtt#connected DO Subscribe2 Occupied, zigbee2mqtt/ikea-motion, occupancy ENDON
	ON Event#Occupied DO Var13 %value%  ENDON
	ON Event#GetVar13 DO Var13 ENDON
```
to this rule, which seems to work:
```
Rule2
	ON mqtt#connected DO Subscribe2 Occupied, home/omgble/BTtoMQTT/582D343996FF, hum ENDON
	ON Event#Occupied DO Var13 %value%  ENDON
	ON Event#GetVar13 DO Var13 ENDON
```
Now I see log messages:
```
11:57:47.289 RUL: EVENT#OCCUPIED performs "Var13 59.2"
11:57:47.291 SRC: Rule
11:57:47.294 CMD: Grp 0, Cmd 'VAR', Idx 13, Len 4, Pld 59, Data '59.2'
11:57:47.301 MQT: stat/tasmota_1C947C/RESULT = {"Var13":"59.2"}
11:57:49.748 RUL: EVENT#OCCUPIED performs "Var13 59.2"
11:57:49.750 SRC: Rule
11:57:49.753 CMD: Grp 0, Cmd 'VAR', Idx 13, Len 4, Pld 59, Data '59.2'
11:57:49.759 MQT: stat/tasmota_1C947C/RESULT = {"Var13":"59.2"}
```

The only thing different between the two rules is `zigbee2mqtt/ikea-motion, occupancy` vs `home/omgble/BTtoMQTT/582D343996FF, hum`.

How do I debug this further to figure out exactly what's going wrong?

**Comment by barbudor** [comment]
Why are you using `Subscribe2` vs `Subscribe`

 I'm using `Subscribe`
Using Subscribe2 means you are looking to sub topics of `zigbee2mqtt/ikea-motion/#` and won't be triggerred for anything published on `zigbee2mqtt/ikea-motion`

**Reply by barbudor** [reply]
As you haven't documented anywhere the exact topic and message that you are receiving...

**Reply by cmangla** [reply]
The exact topic is `zigbee2mqtt/ikea-motion`, and an example message is:
```json
{
"battery":87,
"illuminance_above_threshold":false,
"linkquality":51,
"occupancy":false,
"requested_brightness_level":76,
"requested_brightness_percent":30,
"update":{"installed_version":537011747,"latest_version":537011747,"state":"idle"},
"update_available":false
}
```

**Reply by sfromis** [reply]
Well, I think that `Subscribe2` actually is the better option. The [docs text](https://tasmota.github.io/docs/Commands/#subscribe) looks misleading, I only see the `/#` appended when using `Subscribe` and *not* when using `Subscribe2`.

And it works for me with subscribe2, but interestingly not when the payload is quite long (in this case 259 bytes). I made it work by cutting the exact same payload down to:
`{"battery":87,"illuminance_above_threshold":false,"linkquality":83,"occupancy":true,"requested_brightness_level":76}`

**Reply by cmangla** [reply]
>  I only see the `/#` appended when using `Subscribe` and _not_ when using `Subscribe2`.

That is indeed why I use `Subscribe2`.

> not when the payload is quite long

@sfromis could you point me to the source code where the received MQTT message is parsed/decoded ?

**Reply by sfromis** [reply]
Just scroll a bit back in the source link I already posted.

**Reply by cmangla** [reply, marked as answer]
@sfromis That was a very good hint, because near the top of that file is the line:
```
#define RULE_MAX_MQTT_EVENTSZ   256
```
I set this to 1024 in `user_config_override.h` and now the motion sensor MQTT messages are getting parsed! See:
```
13:11:49.289 RUL: EVENT#OCCUPIED performs "Var13 TRUE"
13:11:49.291 SRC: Rule
13:11:49.293 CMD: Grp 0, Cmd 'VAR', Idx 13, Len 4, Pld 1, Data 'TRUE'
13:11:49.299 MQT: stat/tasmota_1C947C/RESULT = {"Var13":"TRUE"}
```

Thanks very much for your help!

**Reply by cmangla** [reply]
I shall soon raise a pull request to update the docs with this.

**Reply by cmangla** [reply]
See https://github.com/tasmota/docs/pull/1254
